Outline in this Notebook (**`01_data_loading.ipynb`**):

#### **Sections**
1. **Introduction**
   - Brief explanation of the datasets being loaded.

2. **Library Imports**
   - Import necessary libraries.

3. **Configuration Setup**
   - Load `config.yaml`.

4. **Dataset Loading**
   - Use the `data_loader.py` utility to load raw datasets.

5. **Initial Inspection**
   - Display basic stats and preview data.

6. **Save Cleaned Data**
   - Save cleaned datasets to `datasets/processed/`.



### **1. `Introduction`**
This notebook focuses on loading the raw datasets for the Dubai Real Estate Prediction project. We will:
1. Load datasets using `utils/data_loader.py`.
2. Inspect the structure and contents of the datasets.
3. Perform initial cleaning where necessary.
4. Save cleaned datasets to `datasets/processed/`.


### **2. `Library Imports`**

In [1]:
# Import necessary libraries
import os
import pandas as pd
import yaml
import sys

# Display settings
pd.set_option('display.max_colwidth', 50)  # Limit column width for better readability

# Import the data_loader module
import sys
import os

# Add the utils folder to sys.path to import data_loader
sys.path.append("../utils")

# Import the data_loader module
import data_loader
import data_processor
import seaborn as sns; sns.set_theme()


### **3. `Load Configuration`**

In [2]:
from data_loader import load_dataset

# Load configuration
config_path = "../configs/config.yaml"
with open(config_path, "r") as file:
    config = yaml.safe_load(file)

# Print paths and file structure for reference
raw_data_path = config["paths"]["raw_data"]
processed_data_path = config["paths"]["processed_data"]

print(f"Raw Data Path: {raw_data_path}")
print(f"Processed Data Path: {processed_data_path}")


Raw Data Path: ../datasets/raw
Processed Data Path: ../datasets/processed


### **4. `Dataset Loading and Inspection`**

- Expecting 7 Folders and 15 datasets as supplied and retrieved

#### 4.1 `rents.csv`

In [3]:
rents_df = data_loader.inspect_data("01_real_estate_rents", delimiter=";", show_progress=False,loading_method='parallel')

Loading 01_real_estate_rents from ../datasets/raw/Rents & Transactions/rents.csv...


Loading files in parallel: 100%|██████████| 1/1 [01:13<00:00, 73.78s/it]


Shape of 01_real_estate_rents: (5475131, 22)
Columns in 01_real_estate_rents:
Index(['Ejari Contract Number', 'Registration Date', 'Start Date', 'End Date',
       'Property ID', 'Version', 'Area', 'Contract Amount', 'Annual Amount',
       'Is Free Hold?', 'Property Size (sq.m)', 'Property Type',
       'Property Sub Type', 'Number of Rooms', 'Usage', 'Nearest Metro',
       'Nearest Mall', 'Nearest Landmark', 'Parking', 'No of Units',
       'Master Project', 'Project'],
      dtype='object')


- Original state of rents.csv

In [4]:
optimized_rents_df = data_processor.optimize_dataframe_memory(rents_df)
data_processor.display_dataframe_info(optimized_rents_df)

Original memory usage: 918.98 MB
Optimized memory usage: 346.39 MB
Shape: (5475131, 22)
Dtypes: ['Ejari Contract Number:int64', 'Registration Date:object', 'Start Date:category', 'End Date:category', 'Property ID:int64', 'Version:category', 'Area:category', 'Contract Amount:category', 'Annual Amount:category', 'Is Free Hold?:category', 'Property Size (sq.m):category', 'Property Type:category', 'Property Sub Type:category', 'Number of Rooms:float32', 'Usage:category', 'Nearest Metro:category', 'Nearest Mall:category', 'Nearest Landmark:category', 'Parking:float32', 'No of Units:int16', 'Master Project:category', 'Project:category']
Missings (%): ['Ejari Contract Number:0.00%', 'Registration Date:0.00%', 'Start Date:0.00%', 'End Date:0.00%', 'Property ID:0.00%', 'Version:0.00%', 'Area:0.01%', 'Contract Amount:0.00%', 'Annual Amount:0.00%', 'Is Free Hold?:0.00%', 'Property Size (sq.m):0.23%', 'Property Type:0.00%', 'Property Sub Type:0.73%', 'Number of Rooms:96.53%', 'Usage:1.01%', 'Neare

Ejari Contract Number    Registration Date           Start Date  \
0        120210425000238  25/04/2021 09:36:31  13/04/2021 00:00:00   
1        120210526001582  26/05/2021 12:12:22  01/06/2021 00:00:00   

              End Date  Property ID Version                          Area  \
0  12/04/2026 00:00:00   1058329686     New  Dubai Investment Park Second   
1  30/11/2025 00:00:00     99539298     New                Um Nahad Third   

  Contract Amount Annual Amount Is Free Hold?  ... Property Sub Type  \
0           25000          5000     Free Hold  ...           Factory   
1        16000000       4000000     Free Hold  ...          Building   

  Number of Rooms        Usage  Nearest Metro Nearest Mall Nearest Landmark  \
0             NaN   Industrial            NaN          NaN   Expo 2020 Site   
1             NaN  Residential            NaN          NaN              NaN   

  Parking No of Units  Master Project  Project  
0     NaN           1             NaN      NaN  
1     NaN           1             NaN      NaN  

[2 rows x 22 columns]

***Initial Processing Summary for `rents.csv`***


---

**2. Data Types Overview**
**Rental Transactions Schema** table with the missing percentages indicated and notes about data compression from `String` to `Category` where applicable:

| **Field Name**              | **Description**                                      | **Optimized Data Type** | **Notes**                              |
|-----------------------------|----------------------------------------------------|-------------------------|---------------------------------------|
| Ejari Contract Number       | Unique identifier for the rental contract           | `int64`                | No missing values                     |
| Registration Date           | Date of contract registration                       | `object`               | Convertible to datetime; no missing   |
| Start Date                  | Start date of the rental agreement                  | `category`             | Converted from `String`; no missing   |
| End Date                    | End date of the rental agreement                    | `category`             | Converted from `String`; no missing   |
| Property ID                 | Unique identifier for the property                  | `int64`                | No missing values                     |
| Version                     | Version of the contract                             | `category`             | Converted from `String`; no missing   |
| Area                        | Area of Dubai where the property is located         | `category`             | Converted from `String`; 0.01% missing |
| Contract Amount             | Total contract amount in AED                        | `category`             | Converted from `String`; no missing   |
| Annual Amount               | Annual rental amount in AED                         | `category`             | Converted from `String`; no missing   |
| Is Free Hold?               | Indicates if the property is freehold               | `category`             | Converted from `String`; no missing   |
| Property Size (sq.m)        | Area of the property in square meters               | `category`             | Converted from `String`; 0.23% missing |
| Property Type               | General type of property (Building, Land, Unit)     | `category`             | Converted from `String`; no missing   |
| Property Sub Type           | Specific subtype of property (e.g., Workshop, Studio) | `category`           | Converted from `String`; 0.73% missing |
| Number of Rooms             | Number of rooms in the property                     | `float32`             | 96.53% missing                        |
| Usage                       | Usage of the property (Commercial, Industrial, Residential) | `category`     | Converted from `String`; 1.01% missing |
| Nearest Metro               | Name of the nearest metro                           | `category`             | Converted from `String`; 10.27% missing |
| Nearest Mall                | Name of the nearest mall                            | `category`             | Converted from `String`; 10.94% missing |
| Nearest Landmark            | Name of the nearest landmark                        | `category`             | Converted from `String`; 6.33% missing |
| Parking                     | Number of parking spaces                            | `float32`             | 96.93% missing                        |
| No of Units                 | Number of units in the property                     | `int16`               | No missing values                     |
| Master Project              | Name of the master project                          | `category`             | Converted from `String`; 100% missing |
| Project                     | Name of the project                                 | `category`             | Converted from `String`; 85.49% missing |


---

 **3. Missing Values**
- **Total Missing Values (%):** 18.57% of the data.
- **Key Columns with High Missing Percentages:**
  - `Number of Rooms`: 96.53% missing  
  - `Parking`: 96.93% missing  
  - `Master Project`: 100% missing (may require exclusion)  
  - `Project`: 85.49% missing  

---

 **4. Duplicates**
- **Total Duplicates:** 6,398 rows (0.12% of total rows).  

---

**5. Initial Insights**
1. **Memory Optimization** significantly reduces dataset size, making it more efficient for analysis.  
2. High missing percentages in some columns (e.g., `Number of Rooms`, `Parking`, `Master Project`) suggest these fields may not be useful for modeling or analysis without imputation or further validation.  
3. Duplicates should be reviewed for removal, ensuring they don’t bias results.  
4. Categorical conversion for appropriate fields helps in reducing memory while preserving information.  

---

**6. Further Actions**
1. **Handle Missing Data:** 
   - Considering removal of columns like `Master Project` with 100% missing values. Others with over 90% can also be removed
   - Exploreing imputation strategies for other critical columns especially compulsory ones required for modeling.  
2. **Remove Duplicates:** Dropping duplicate rows to maintain data integrity.  
3. **Datetime Parsing:** Converting `Registration Date`, `Start Date`, and `End Date` into proper datetime objects.  
4. **Further Validation:** Checking for inconsistencies in categorical values and ensure data quality.

In [ ]:
# sns.heatmap(rents_df.isnull(), cbar=False, cmap='viridis')

<Axes: >

In [6]:
optimized_rents_df[['Usage','Nearest Metro','Nearest Mall','Nearest Landmark']].isnull().mean() * 100

Usage                1.007684
Nearest Metro       10.266184
Nearest Mall        10.938898
Nearest Landmark     6.333273
dtype: float64

<!-- Property Type or/and Property Sub Type
Number of rooms/bedrooms
Property Size (sq.m)
Area/Neighborhood
Nearest Metro
Nearest Mall
Nearest Landmark
Usage (commercial/residential)
Average prices the previous month/week (for the same kind of property) -->


In [7]:
# Specify columns to format
datetime_cols = ["Registration Date", "Start Date", "End Date"]
numeric_cols = ["Contract Amount", "Annual Amount", "Property Size (sq.m)", "Number of Rooms"]

processed_rents_df = data_processor.format_and_clean_table(df=optimized_rents_df, datetime_columns=datetime_cols, numeric_columns=numeric_cols, missing_threshold=0.8)
processed_rents_df = data_processor.optimize_dataframe_memory(processed_rents_df)
data_processor.display_dataframe_info(processed_rents_df)

Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
Index: 5380162 entries, 0 to 5475130
Data columns (total 18 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   Ejari Contract Number  int64         
 1   Registration Date      datetime64[ns]
 2   Start Date             category      
 3   End Date               datetime64[ns]
 4   Property ID            int64         
 5   Version                category      
 6   Area                   category      
 7   Contract Amount        float64       
 8   Annual Amount          float64       
 9   Is Free Hold?          category      
 10  Property Size (sq.m)   float64       
 11  Property Type          category      
 12  Property Sub Type      category      
 13  Usage                  category      
 14  Nearest Metro          category      
 15  Nearest Mall           category      
 16  Nearest Landmark       category      
 17  No of Units            int64     

Ejari Contract Number   Registration Date Start Date   End Date  \
0        120210425000238 2021-04-25 09:36:31 2021-04-13 2026-04-12   
1        120210526001582 2021-05-26 12:12:22 2021-06-01 2025-11-30   

   Property ID Version                          Area  Contract Amount  \
0   1058329686     New  Dubai Investment Park Second          25000.0   
1     99539298     New                Um Nahad Third       16000000.0   

   Annual Amount Is Free Hold?  Property Size (sq.m) Property Type  \
0         5000.0     Free Hold                   NaN      Building   
1      4000000.0     Free Hold                6792.0      Building   

  Property Sub Type        Usage Nearest Metro Nearest Mall Nearest Landmark  \
0           Factory   Industrial           NaN          NaN   Expo 2020 Site   
1          Building  Residential           NaN          NaN              NaN   

   No of Units  
0            1  
1            1

In [8]:
print("Cleaned Data Preview:")
processed_rents_df.isnull().mean() * 100

Cleaned Data Preview:


Ejari Contract Number     0.000000
Registration Date         0.000000
Start Date                0.000000
End Date                  0.000037
Property ID               0.000000
Version                   0.000000
Area                      0.005911
Contract Amount           2.789674
Annual Amount             5.823691
Is Free Hold?             0.000000
Property Size (sq.m)     50.145367
Property Type             0.000000
Property Sub Type         0.741929
Usage                     1.024226
Nearest Metro             9.545475
Nearest Mall             10.229785
Nearest Landmark          6.427706
No of Units               0.000000
dtype: float64

In [9]:
invalid_dates = processed_rents_df[pd.to_datetime(processed_rents_df["Start Date"], format="%d/%m/%Y %H:%M:%S", errors='coerce').isna()]
print("Invalid date entries in 'Start Date':")
print(invalid_dates)


Invalid date entries in 'Start Date':
Empty DataFrame
Columns: [Ejari Contract Number, Registration Date, Start Date, End Date, Property ID, Version, Area, Contract Amount, Annual Amount, Is Free Hold?, Property Size (sq.m), Property Type, Property Sub Type, Usage, Nearest Metro, Nearest Mall, Nearest Landmark, No of Units]
Index: []


#### 4.2 `transactions.csv`

In [10]:
transactions_df = data_loader.inspect_data("02_real_estate_transactions", delimiter=";", show_progress=False,loading_method='parallel')

Loading 02_real_estate_transactions from ../datasets/raw/Rents & Transactions/transactions.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:11<00:00, 11.12s/it]

Shape of 02_real_estate_transactions: (995975, 23)
Columns in 02_real_estate_transactions:
Index(['Transaction Number', 'Transaction Date', 'Property ID',
       'Transaction Type', 'Transaction sub type', 'Registration type',
       'Is Free Hold?', 'Usage', 'Area', 'Property Type', 'Property Sub Type',
       'Amount', 'Transaction Size (sq.m)', 'Property Size (sq.m)', 'Room(s)',
       'Parking', 'Nearest Metro', 'Nearest Mall', 'Nearest Landmark',
       'No. of Buyer', 'No. of Seller', 'Master Project', 'Project'],
      dtype='object')


In [11]:
optimized_transactions_df = data_processor.optimize_dataframe_memory(transactions_df)
data_processor.display_dataframe_info(optimized_transactions_df)

Original memory usage: 174.77 MB
Optimized memory usage: 71.09 MB
Shape: (995975, 23)
Dtypes: ['Transaction Number:object', 'Transaction Date:object', 'Property ID:int32', 'Transaction Type:category', 'Transaction sub type:category', 'Registration type:category', 'Is Free Hold?:category', 'Usage:category', 'Area:category', 'Property Type:category', 'Property Sub Type:category', 'Amount:float64', 'Transaction Size (sq.m):float64', 'Property Size (sq.m):float64', 'Room(s):category', 'Parking:category', 'Nearest Metro:category', 'Nearest Mall:category', 'Nearest Landmark:category', 'No. of Buyer:float32', 'No. of Seller:float32', 'Master Project:category', 'Project:category']
Missings (%): ['Transaction Number:0.00%', 'Transaction Date:0.00%', 'Property ID:0.00%', 'Transaction Type:0.00%', 'Transaction sub type:0.00%', 'Registration type:0.00%', 'Is Free Hold?:0.00%', 'Usage:0.00%', 'Area:0.00%', 'Property Type:0.00%', 'Property Sub Type:2.23%', 'Amount:0.00%', 'Transaction Size (sq.m):12

Transaction Number  Transaction Date  Property ID Transaction Type  \
0       13-7700-2022  02/06/2022 16:18     44581083         Mortgage   
1       13-7699-2022  02/06/2022 16:14     44014284         Mortgage   

    Transaction sub type Registration type Is Free Hold?        Usage  \
0  Mortgage Registration             Ready     Free Hold  Residential   
1  Mortgage Registration             Ready     Free Hold  Residential   

           Area Property Type  ... Property Size (sq.m)  Room(s)  Parking  \
0  AKOYA OXYGEN          Land  ...                174.8      NaN      NaN   
1  AKOYA OXYGEN          Land  ...                174.8      NaN      NaN   

   Nearest Metro Nearest Mall      Nearest Landmark No. of Buyer  \
0            NaN          NaN  Dubai Cycling Course          0.0   
1            NaN          NaN  Dubai Cycling Course          0.0   

  No. of Seller Master Project                    Project  
0           0.0            NaN   DAMAC HILLS (2) - ODORA   
1           0.0            NaN  DAMAC HILLS (2) - ZINNIA   

[2 rows x 23 columns]

**Transactions Data Schema** table with missing percentages and data compression details:

| **Field Name**              | **Description**                                           | **Optimized Data Type** | **Notes**                              |
|-----------------------------|----------------------------------------------------------|-------------------------|---------------------------------------|
| Transaction Number          | Unique identifier for the transaction                     | `object`               | No missing values                     |
| Transaction Date            | Date and time of the transaction                          | `object`               | Convertible to datetime; no missing   |
| Property ID                 | Unique identifier for the property                        | `int32`                | No missing values                     |
| Transaction Type            | Type of transaction (Sales, Gifts, Mortgage)             | `category`             | Converted from `String`; no missing   |
| Transaction sub type         | Subtype of the transaction (e.g., Delayed Sell, Mortgage) | `category`             | Converted from `String`; no missing   |
| Registration type           | Type of registration (Ready, Off-Plan)                   | `category`             | Converted from `String`; no missing   |
| Is Free Hold?               | Indicates if the property is freehold                    | `category`             | Converted from `String`; no missing   |
| Usage                       | Usage of the property (Commercial, Residential)          | `category`             | Converted from `String`; no missing   |
| Area                        | Area of Dubai where the property is located              | `category`             | Converted from `String`; no missing   |
| Property Type               | General type of property (Unit, Land, Building)          | `category`             | Converted from `String`; no missing   |
| Property Sub Type           | Specific subtype of property (e.g., Flat, Villa, Shop)   | `category`             | Converted from `String`; 2.23% missing |
| Amount                      | Transaction amount in AED                                 | `float64`              | No missing values                     |
| Transaction Size (sq.m)     | Area of the transaction in square meters                 | `float64`              | 12.62% missing                        |
| Property Size (sq.m)        | Area of the property in square meters                    | `float64`              | No missing values                     |
| Room(s)                     | Property layout (e.g., Studio, 1 B/R, 5 B/R)             | `category`             | Converted from `String`; 23.88% missing |
| Parking                     | Number of parking spaces                                 | `category`             | Converted from `String`; 49.53% missing |
| Nearest Metro               | Name of the nearest metro                                | `category`             | Converted from `String`; 21.70% missing |
| Nearest Mall                | Name of the nearest mall                                 | `category`             | Converted from `String`; 22.10% missing |
| Nearest Landmark            | Name of the nearest landmark                             | `category`             | Converted from `String`; 10.38% missing |
| No. of Buyer                | Number of buyers in the transaction                      | `float32`              | 0.02% missing                         |
| No. of Seller               | Number of sellers in the transaction                    | `float32`              | 0.02% missing                         |
| Master Project              | Name of the master project                               | `category`             | Converted from `String`; 99.97% missing |
| Project                     | Name of the project                                      | `category`             | Converted from `String`; 40.89% missing |

---

1. **Missing Data Indicators**:
   - Missing percentages are explicitly listed for each field.
   - Fields like `Master Project` and `Parking` have significant missing data and might need to be excluded or imputed.
   
2. **Compression Details**:
   - Several fields (e.g., `Transaction Type`, `Usage`, `Nearest Metro`) were optimized from `String` to `Category`, significantly reducing memory usage.

3. **Key Insights**:
   - Memory usage reduced from **174.77 MB** to **71.09 MB** (60% improvement).
   - Duplicates: **0**, indicating no redundancy.
   - Total missing data: **12.32%**, concentrated in fields like `Transaction Size`, `Room(s)`, and `Parking`.


In [12]:
# transactions_df['Room(s)'].unique()
transactions_df[['Transaction Number','Transaction Date']]

Transaction Number  Transaction Date
0            13-7700-2022  02/06/2022 16:18
1            13-7699-2022  02/06/2022 16:14
2             14-563-2022  02/06/2022 15:42
3           11-12317-2022  02/06/2022 15:22
4           11-12316-2022  02/06/2022 15:07
...                   ...               ...
995970       13-1699-1996  14/02/1996 00:00
995971       13-1701-1996  24/01/1996 00:00
995972       13-1742-1995  26/10/1995 00:00
995973       13-1746-1995  18/10/1995 00:00
995974       13-1744-1995  07/03/1995 00:00

[995975 rows x 2 columns]

In [13]:
# Specify columns to format
# datetime_cols = ["Transaction Date"]
# numeric_cols = ["Transaction Number"]

processed_transactions_df = data_processor.format_and_clean_table(df=optimized_transactions_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 995975 entries, 0 to 995974
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Transaction Number       995975 non-null  object  
 1   Transaction Date         995975 non-null  object  
 2   Property ID              995975 non-null  int64   
 3   Transaction Type         995975 non-null  category
 4   Transaction sub type     995975 non-null  category
 5   Registration type        995975 non-null  category
 6   Is Free Hold?            995975 non-null  category
 7   Usage                    995975 non-null  category
 8   Area                     995975 non-null  category
 9   Property Type            995975 non-null  category
 10  Property Sub Type        973723 non-null  category
 11  Amount                   995975 non-null  float64 
 12  Transaction Size (sq.m)  870248 non-null  float64 
 13  Propert

- Consumer Price Index

#### 4.3 `Consumer_Price_Index_Annually.csv`

In [20]:
cpi_annual_df = data_loader.inspect_data("03_cpi_annual", delimiter=",", show_progress=False,loading_method='parallel')


Loading 03_cpi_annual from ../datasets/raw/Consumer Price Index/Consumer_Price_Index_Annually.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:01<00:00,  1.04s/it]

Shape of 03_cpi_annual: (369, 12)
Columns in 03_cpi_annual:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'CPI_DIV',
       'SOURCE_DETAIL', 'TIME_PERIOD', 'OBS_VALUE', 'DECIMALS', 'OBS_STATUS',
       'UNIT_MULT'],
      dtype='object')


In [21]:
cpi_annual_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
CPI_DIV            0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS           0.0
OBS_STATUS       100.0
UNIT_MULT        100.0
dtype: float64

In [22]:
cpi_annual_df

DATAFLOW       MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0    FCSA:DF_CPI_ANN(3.2.0)   CPI_INDEX21         INDX       AE    A     FUR   
1    FCSA:DF_CPI_ANN(3.2.0)   CPI_INDEX21         INDX       AE    A     FUR   
2    FCSA:DF_CPI_ANN(3.2.0)   CPI_INDEX21         INDX       AE    A     FUR   
3    FCSA:DF_CPI_ANN(3.2.0)    CPI_ANNCHG      PERCENT       AE    A     TEX   
4    FCSA:DF_CPI_ANN(3.2.0)    CPI_ANNCHG      PERCENT       AE    A     TEX   
..                      ...           ...          ...      ...  ...     ...   
364  FCSA:DF_CPI_ANN(3.2.0)    CPI_ANNCHG      PERCENT       AE    A     REC   
365  FCSA:DF_CPI_ANN(3.2.0)  CPI_ANNCHG21      PERCENT       AE    A     EDU   
366  FCSA:DF_CPI_ANN(3.2.0)  CPI_ANNCHG21      PERCENT       AE    A     EDU   
367  FCSA:DF_CPI_ANN(3.2.0)  CPI_ANNCHG21      PERCENT       AE    A     RES   
368  FCSA:DF_CPI_ANN(3.2.0)  CPI_ANNCHG21      PERCENT       AE    A     RES   

    SOURCE_DETAIL  TIME_PERIOD   OBS_VALUE  DECIMALS  OBS_STATUS  UNIT_MULT  
0            FCSA         2021  100.000000         2         NaN        NaN  
1            FCSA         2022  100.020000         2         NaN        NaN  
2            FCSA         2023  103.223874         2         NaN        NaN  
3            FCSA         2010   -4.960000         2         NaN        NaN  
4            FCSA         2011   -1.870000         2         NaN        NaN  
..            ...          ...         ...       ...         ...        ...  
364          FCSA         2021    3.110000         2         NaN        NaN  
365          FCSA         2022    0.950000         2         NaN        NaN  
366          FCSA         2023    1.372461         2         NaN        NaN  
367          FCSA         2022    7.190000         2         NaN        NaN  
368          FCSA         2023    6.625620         2         NaN        NaN  

[369 rows x 12 columns]

In [23]:
# datetime_cols = ["TIME_PERIOD"]
# numeric_cols = ["Transaction Number"]

processed_cpi_annual_df = data_processor.format_and_clean_table(df=cpi_annual_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 369 entries, 0 to 368
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       369 non-null    object 
 1   MEASURE        369 non-null    object 
 2   UNIT_MEASURE   369 non-null    object 
 3   REF_AREA       369 non-null    object 
 4   FREQ           369 non-null    object 
 5   CPI_DIV        369 non-null    object 
 6   SOURCE_DETAIL  369 non-null    object 
 7   TIME_PERIOD    369 non-null    int64  
 8   OBS_VALUE      369 non-null    float64
 9   DECIMALS       369 non-null    int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 29.0+ KB
None


In [24]:
processed_cpi_annual_df.isnull().mean() * 100

DATAFLOW         0.0
MEASURE          0.0
UNIT_MEASURE     0.0
REF_AREA         0.0
FREQ             0.0
CPI_DIV          0.0
SOURCE_DETAIL    0.0
TIME_PERIOD      0.0
OBS_VALUE        0.0
DECIMALS         0.0
dtype: float64

In [25]:
# optimized_cpi_annual_df = data_processor.optimize_dataframe_memory(cpi_annual_df)
data_processor.display_dataframe_info(processed_cpi_annual_df)


Shape: (369, 10)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'CPI_DIV:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:int64', 'OBS_VALUE:float64', 'DECIMALS:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'CPI_DIV:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%', 'DECIMALS:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.03 MB


DATAFLOW      MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0  FCSA:DF_CPI_ANN(3.2.0)  CPI_INDEX21         INDX       AE    A     FUR   
1  FCSA:DF_CPI_ANN(3.2.0)  CPI_INDEX21         INDX       AE    A     FUR   

  SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  DECIMALS  
0          FCSA         2021     100.00         2  
1          FCSA         2022     100.02         2

#### 4.4 `Consumer_Price_Index_Monthly.csv`

In [26]:
cpi_monthly_df = data_loader.inspect_data("04_cpi_monthly", delimiter=",", show_progress=False,loading_method='parallel')


Loading 04_cpi_monthly from ../datasets/raw/Consumer Price Index/Consumer_Price_Index_Monthly.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.12it/s]

Shape of 04_cpi_monthly: (7361, 12)
Columns in 04_cpi_monthly:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'CPI_DIV',
       'SOURCE_DETAIL', 'TIME_PERIOD', 'OBS_VALUE', 'DECIMALS', 'OBS_STATUS',
       'UNIT_MULT'],
      dtype='object')


In [27]:
cpi_monthly_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
CPI_DIV            0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS           0.0
OBS_STATUS       100.0
UNIT_MULT        100.0
dtype: float64

In [28]:
cpi_monthly_df

DATAFLOW     MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0     FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
1     FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
2     FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
3     FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
4     FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
...                  ...         ...          ...      ...  ...     ...   
7356  FCSA:DF_CPI(3.2.0)  CPI_MTHCHG      PERCENT       AE    M     INS   
7357  FCSA:DF_CPI(3.2.0)  CPI_MTHCHG      PERCENT       AE    M     INS   
7358  FCSA:DF_CPI(3.2.0)  CPI_MTHCHG      PERCENT       AE    M     INS   
7359  FCSA:DF_CPI(3.2.0)  CPI_MTHCHG      PERCENT       AE    M     INS   
7360  FCSA:DF_CPI(3.2.0)  CPI_MTHCHG      PERCENT       AE    M     INS   

     SOURCE_DETAIL TIME_PERIOD  OBS_VALUE  DECIMALS  OBS_STATUS  UNIT_MULT  
0             FCSA     2009-01   7.250000         2         NaN        NaN  
1             FCSA     2009-02   3.870000         2         NaN        NaN  
2             FCSA     2009-03   4.700000         2         NaN        NaN  
3             FCSA     2009-04   1.900000         2         NaN        NaN  
4             FCSA     2009-05   5.250000         2         NaN        NaN  
...            ...         ...        ...       ...         ...        ...  
7356          FCSA     2024-03  -1.114635         2         NaN        NaN  
7357          FCSA     2024-04  -0.022338         2         NaN        NaN  
7358          FCSA     2024-05  -0.005594         2         NaN        NaN  
7359          FCSA     2024-06   2.137716         2         NaN        NaN  
7360          FCSA     2024-07   0.299637         2         NaN        NaN  

[7361 rows x 12 columns]

In [29]:
processed_cpi_monthly_df = data_processor.format_and_clean_table(df=cpi_monthly_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7361 entries, 0 to 7360
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       7361 non-null   object 
 1   MEASURE        7361 non-null   object 
 2   UNIT_MEASURE   7361 non-null   object 
 3   REF_AREA       7361 non-null   object 
 4   FREQ           7361 non-null   object 
 5   CPI_DIV        7361 non-null   object 
 6   SOURCE_DETAIL  7361 non-null   object 
 7   TIME_PERIOD    7361 non-null   object 
 8   OBS_VALUE      7361 non-null   float64
 9   DECIMALS       7361 non-null   int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 575.2+ KB
None


In [30]:
# optimized_cpi_monthly_df = data_processor.optimize_dataframe_memory(cpi_monthly_df)
data_processor.display_dataframe_info(processed_cpi_monthly_df)


Shape: (7361, 10)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'CPI_DIV:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:object', 'OBS_VALUE:float64', 'DECIMALS:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'CPI_DIV:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%', 'DECIMALS:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.56 MB


DATAFLOW     MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0  FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   
1  FCSA:DF_CPI(3.2.0)  CPI_ANNCHG      PERCENT       AE    M     FUR   

  SOURCE_DETAIL TIME_PERIOD  OBS_VALUE  DECIMALS  
0          FCSA     2009-01       7.25         2  
1          FCSA     2009-02       3.87         2

#### 4.5 `Consumer_Price_Index_Quarterly.csv`

In [31]:
cpi_quarterly_df = data_loader.inspect_data("05_cpi_quarterly", delimiter=",", show_progress=False,loading_method='parallel')


Loading 05_cpi_quarterly from ../datasets/raw/Consumer Price Index/Consumer_Price_Index_Quarterly.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.10it/s]

Shape of 05_cpi_quarterly: (112, 12)
Columns in 05_cpi_quarterly:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'CPI_DIV',
       'SOURCE_DETAIL', 'TIME_PERIOD', 'OBS_VALUE', 'DECIMALS', 'OBS_STATUS',
       'UNIT_MULT'],
      dtype='object')


In [32]:
cpi_quarterly_df

DATAFLOW    MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0    FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
1    FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
2    FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
3    FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
4    FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
..                    ...        ...          ...      ...  ...     ...   
107  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     TEX   
108  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     TEX   
109  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     TEX   
110  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     TEX   
111  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     TEX   

    SOURCE_DETAIL TIME_PERIOD  OBS_VALUE  DECIMALS  OBS_STATUS  UNIT_MULT  
0            FCSA     2022-Q1  -1.851788         2         NaN        NaN  
1            FCSA     2022-Q2  -1.001258         2         NaN        NaN  
2            FCSA     2022-Q3   0.119038         2         NaN        NaN  
3            FCSA     2022-Q4   1.605925         2         NaN        NaN  
4            FCSA     2023-Q1   2.451617         2         NaN        NaN  
..            ...         ...        ...       ...         ...        ...  
107          FCSA     2022-Q4   4.129407         2         NaN        NaN  
108          FCSA     2023-Q1   4.650506         2         NaN        NaN  
109          FCSA     2023-Q2   3.903968         2         NaN        NaN  
110          FCSA     2023-Q3   2.964095         2         NaN        NaN  
111          FCSA     2023-Q4   1.252795         2         NaN        NaN  

[112 rows x 12 columns]

In [33]:
cpi_quarterly_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
CPI_DIV            0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS           0.0
OBS_STATUS       100.0
UNIT_MULT        100.0
dtype: float64

In [34]:
processed_cpi_quarterly_df = data_processor.format_and_clean_table(df=cpi_quarterly_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112 entries, 0 to 111
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       112 non-null    object 
 1   MEASURE        112 non-null    object 
 2   UNIT_MEASURE   112 non-null    object 
 3   REF_AREA       112 non-null    object 
 4   FREQ           112 non-null    object 
 5   CPI_DIV        112 non-null    object 
 6   SOURCE_DETAIL  112 non-null    object 
 7   TIME_PERIOD    112 non-null    object 
 8   OBS_VALUE      112 non-null    float64
 9   DECIMALS       112 non-null    int64  
dtypes: float64(1), int64(1), object(8)
memory usage: 8.9+ KB
None


In [35]:
# optimized_cpi_monthly_df = data_processor.optimize_dataframe_memory(cpi_monthly_df)
data_processor.display_dataframe_info(processed_cpi_quarterly_df)


Shape: (112, 10)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'CPI_DIV:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:object', 'OBS_VALUE:float64', 'DECIMALS:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'CPI_DIV:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%', 'DECIMALS:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.01 MB


DATAFLOW    MEASURE UNIT_MEASURE REF_AREA FREQ CPI_DIV  \
0  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   
1  FCSA:DF_CPI_Q(3.2.0)  CPI_QCHNG      PERCENT       AE    Q     HOU   

  SOURCE_DETAIL TIME_PERIOD  OBS_VALUE  DECIMALS  
0          FCSA     2022-Q1  -1.851788         2  
1          FCSA     2022-Q2  -1.001258         2

- Population

#### 4.6 `Population_Estimates_and_Growth_by_Gender.csv`

In [36]:
pop_estimates_df = data_loader.inspect_data("06_population_estimates", delimiter=",", show_progress=False,loading_method='parallel')


Loading 06_population_estimates from ../datasets/raw/Population/Population_Estimates_and_Growth_by_Gender.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.09it/s]

Shape of 06_population_estimates: (162, 11)
Columns in 06_population_estimates:
Index(['DATAFLOW', 'MEASURE', 'POP_IND', 'UNIT_MEASURE', 'REF_AREA', 'FREQ',
       'GENDER', 'SOURCE_DETAIL', 'TIME_PERIOD', 'OBS_VALUE', 'DECIMALS'],
      dtype='object')


In [37]:
pop_estimates_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
POP_IND            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
GENDER             0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS         100.0
dtype: float64

In [38]:
pop_estimates_df

DATAFLOW MEASURE POP_IND UNIT_MEASURE REF_AREA FREQ GENDER  \
0    FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
1    FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
2    FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
3    FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
4    FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
..                  ...     ...     ...          ...      ...  ...    ...   
157  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A     _T   
158  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A     _T   
159  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A     _T   
160  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A     _T   
161  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A     _T   

    SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  DECIMALS  
0            FCSC         1970     149195       NaN  
1            FCSC         1971     181473       NaN  
2            FCSC         1972     221774       NaN  
3            FCSC         1973     268398       NaN  
4            FCSC         1974     322544       NaN  
..            ...          ...        ...       ...  
157          FCSC         2019    9512778       NaN  
158          FCSC         2020    9289297       NaN  
159          FCSC         2021    9861007       NaN  
160          FCSC         2022   10288946       NaN  
161          FCSC         2023   10678556       NaN  

[162 rows x 11 columns]

In [39]:
processed_pop_estimates_df = data_processor.format_and_clean_table(df=pop_estimates_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   DATAFLOW       162 non-null    object
 1   MEASURE        162 non-null    object
 2   POP_IND        162 non-null    object
 3   UNIT_MEASURE   162 non-null    object
 4   REF_AREA       162 non-null    object
 5   FREQ           162 non-null    object
 6   GENDER         162 non-null    object
 7   SOURCE_DETAIL  162 non-null    object
 8   TIME_PERIOD    162 non-null    int64 
 9   OBS_VALUE      162 non-null    int64 
dtypes: int64(2), object(8)
memory usage: 12.8+ KB
None


In [40]:
# optimized_cpi_monthly_df = data_processor.optimize_dataframe_memory(cpi_monthly_df)
data_processor.display_dataframe_info(processed_pop_estimates_df)


Shape: (162, 10)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'POP_IND:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'GENDER:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:int64', 'OBS_VALUE:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'POP_IND:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'GENDER:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.01 MB


DATAFLOW MEASURE POP_IND UNIT_MEASURE REF_AREA FREQ GENDER  \
0  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   
1  FCSA:DF_POP(2.7.0)     POP      _Z           PS       AE    A      M   

  SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  
0          FCSC         1970     149195  
1          FCSC         1971     181473

#### 4.7 `Population_Indicators.csv`

In [41]:
pop_indicators_df = data_loader.inspect_data("07_population_indicators", delimiter=",", show_progress=False,loading_method='parallel')


Loading 07_population_indicators from ../datasets/raw/Population/Population_Indicators.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.08it/s]

Shape of 07_population_indicators: (18, 11)
Columns in 07_population_indicators:
Index(['DATAFLOW', 'MEASURE', 'POP_IND', 'UNIT_MEASURE', 'REF_AREA', 'FREQ',
       'GENDER', 'SOURCE_DETAIL', 'TIME_PERIOD', 'OBS_VALUE', 'DECIMALS'],
      dtype='object')


In [42]:
pop_indicators_df.isnull().mean() * 100

DATAFLOW         0.0
MEASURE          0.0
POP_IND          0.0
UNIT_MEASURE     0.0
REF_AREA         0.0
FREQ             0.0
GENDER           0.0
SOURCE_DETAIL    0.0
TIME_PERIOD      0.0
OBS_VALUE        0.0
DECIMALS         0.0
dtype: float64

In [43]:
data_processor.display_dataframe_info(pop_indicators_df)

Shape: (18, 11)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'POP_IND:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'GENDER:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:int64', 'OBS_VALUE:float64', 'DECIMALS:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'POP_IND:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'GENDER:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%', 'DECIMALS:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.00 MB


DATAFLOW MEASURE POP_IND UNIT_MEASURE REF_AREA FREQ GENDER  \
0  FCSA:DF_POP_IND(2.7.0)  POPIND     MED        YEARS       AE    A     _Z   
1  FCSA:DF_POP_IND(2.7.0)  POPIND     MED        YEARS       AE    A     _Z   

  SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  DECIMALS  
0          FCSC         2019       31.3         1  
1          FCSC         2020       32.8         1

- Tourism

#### 4.8 `Guests_by_Hotel_Type_by_Region.csv`

In [44]:
tourism_regions_df = data_loader.inspect_data("08_tourism_regions", delimiter=",", show_progress=False,loading_method='parallel')


Loading 08_tourism_regions from ../datasets/raw/Tourism/Guests_by_Hotel_Type_by_Region.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.11it/s]

Shape of 08_tourism_regions: (664, 15)
Columns in 08_tourism_regions:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'H_TYPE',
       'H_INDICATOR', 'GUEST_REGION', 'SOURCE_DETAIL', 'TIME_PERIOD',
       'OBS_VALUE', 'DECIMALS', 'OBS_STATUS', 'OBS_COMMENT', 'UNIT_MULT'],
      dtype='object')


In [45]:
tourism_regions_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
H_TYPE             0.0
H_INDICATOR        0.0
GUEST_REGION       0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS         100.0
OBS_STATUS       100.0
OBS_COMMENT      100.0
UNIT_MULT        100.0
dtype: float64

In [46]:
processed_tourism_regions_df = data_processor.format_and_clean_table(df=tourism_regions_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 664 entries, 0 to 663
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   DATAFLOW       664 non-null    object
 1   MEASURE        664 non-null    object
 2   UNIT_MEASURE   664 non-null    object
 3   REF_AREA       664 non-null    object
 4   FREQ           664 non-null    object
 5   H_TYPE         664 non-null    object
 6   H_INDICATOR    664 non-null    object
 7   GUEST_REGION   664 non-null    object
 8   SOURCE_DETAIL  664 non-null    object
 9   TIME_PERIOD    664 non-null    int64 
 10  OBS_VALUE      664 non-null    int64 
dtypes: int64(2), object(9)
memory usage: 57.2+ KB
None


In [47]:
data_processor.display_dataframe_info(processed_tourism_regions_df)

Shape: (664, 11)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'H_TYPE:object', 'H_INDICATOR:object', 'GUEST_REGION:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:int64', 'OBS_VALUE:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'H_TYPE:0.00%', 'H_INDICATOR:0.00%', 'GUEST_REGION:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.06 MB


DATAFLOW MEASURE UNIT_MEASURE REF_AREA FREQ H_TYPE  \
0  FCSA:DF_GUEST_REGION(4.3.0)       H       NUMBER       AE    A     _Z   
1  FCSA:DF_GUEST_REGION(4.3.0)       H       NUMBER       AE    A     _Z   

  H_INDICATOR GUEST_REGION SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  
0         GHH           OC          FCSC         2015     399751  
1         GHH           OC          FCSC         2016     380717

#### 4.9 `Hotel_Establishments_and_Rooms_by_Rating_Type.csv`

In [48]:
tourism_ratings_df = data_loader.inspect_data("09_tourism_ratings", delimiter=",", show_progress=False,loading_method='parallel')


Loading 09_tourism_ratings from ../datasets/raw/Tourism/Hotel_Establishments_and_Rooms_by_Rating_Type.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.03it/s]

Shape of 09_tourism_ratings: (400, 15)
Columns in 09_tourism_ratings:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'H_TYPE',
       'H_INDICATOR', 'GUEST_REGION', 'SOURCE_DETAIL', 'TIME_PERIOD',
       'OBS_VALUE', 'DECIMALS', 'OBS_STATUS', 'OBS_COMMENT', 'UNIT_MULT'],
      dtype='object')


In [49]:
tourism_ratings_df.isnull().mean() * 100

DATAFLOW           0.0
MEASURE            0.0
UNIT_MEASURE       0.0
REF_AREA           0.0
FREQ               0.0
H_TYPE             0.0
H_INDICATOR        0.0
GUEST_REGION       0.0
SOURCE_DETAIL      0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
DECIMALS         100.0
OBS_STATUS       100.0
OBS_COMMENT      100.0
UNIT_MULT        100.0
dtype: float64

In [50]:
processed_tourism_ratings_df = data_processor.format_and_clean_table(df=tourism_ratings_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   DATAFLOW       400 non-null    object
 1   MEASURE        400 non-null    object
 2   UNIT_MEASURE   400 non-null    object
 3   REF_AREA       400 non-null    object
 4   FREQ           400 non-null    object
 5   H_TYPE         400 non-null    object
 6   H_INDICATOR    400 non-null    object
 7   GUEST_REGION   400 non-null    object
 8   SOURCE_DETAIL  400 non-null    object
 9   TIME_PERIOD    400 non-null    int64 
 10  OBS_VALUE      400 non-null    int64 
dtypes: int64(2), object(9)
memory usage: 34.5+ KB
None


In [51]:
data_processor.display_dataframe_info(processed_tourism_ratings_df)

Shape: (400, 11)
Dtypes: ['DATAFLOW:object', 'MEASURE:object', 'UNIT_MEASURE:object', 'REF_AREA:object', 'FREQ:object', 'H_TYPE:object', 'H_INDICATOR:object', 'GUEST_REGION:object', 'SOURCE_DETAIL:object', 'TIME_PERIOD:int64', 'OBS_VALUE:int64']
Missings (%): ['DATAFLOW:0.00%', 'MEASURE:0.00%', 'UNIT_MEASURE:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'H_TYPE:0.00%', 'H_INDICATOR:0.00%', 'GUEST_REGION:0.00%', 'SOURCE_DETAIL:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.03 MB


DATAFLOW MEASURE UNIT_MEASURE REF_AREA FREQ H_TYPE  \
0  FCSA:DF_HOT_TYPE(4.3.0)       H       NUMBER       AE    A     _T   
1  FCSA:DF_HOT_TYPE(4.3.0)       H       NUMBER       AE    A     _T   

  H_INDICATOR GUEST_REGION SOURCE_DETAIL  TIME_PERIOD  OBS_VALUE  
0         EST           _Z          FCSC         2014       1027  
1         EST           _Z          FCSC         2015       1056

#### 4.10 `Hotel_Establishments_Main_Indicators.csv`

In [52]:
tourism_indicators_df = data_loader.inspect_data("10_tourism_indicators", delimiter=",", show_progress=False,loading_method='parallel')


Loading 10_tourism_indicators from ../datasets/raw/Tourism/Hotel_Establishments_Main_Indicators.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.10it/s]

Shape of 10_tourism_indicators: (201, 15)
Columns in 10_tourism_indicators:
Index(['DATAFLOW', 'MEASURE', 'UNIT_MEASURE', 'REF_AREA', 'FREQ', 'H_TYPE',
       'H_INDICATOR', 'GUEST_REGION', 'SOURCE_DETAIL', 'TIME_PERIOD',
       'OBS_VALUE', 'DECIMALS', 'OBS_STATUS', 'OBS_COMMENT', 'UNIT_MULT'],
      dtype='object')


In [53]:
tourism_indicators_df['DECIMALS'].unique()

array([nan,  2.,  0.])

In [54]:
tourism_indicators_df.isnull().mean() * 100

DATAFLOW           0.000000
MEASURE            0.000000
UNIT_MEASURE       0.000000
REF_AREA           0.000000
FREQ               0.000000
H_TYPE             0.000000
H_INDICATOR        0.000000
GUEST_REGION       0.000000
SOURCE_DETAIL      0.000000
TIME_PERIOD        0.000000
OBS_VALUE          0.000000
DECIMALS          82.089552
OBS_STATUS       100.000000
OBS_COMMENT      100.000000
UNIT_MULT        100.000000
dtype: float64

In [55]:
processed_tourism_indicators_df = data_processor.format_and_clean_table(df=tourism_indicators_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       201 non-null    object 
 1   MEASURE        201 non-null    object 
 2   UNIT_MEASURE   201 non-null    object 
 3   REF_AREA       201 non-null    object 
 4   FREQ           201 non-null    object 
 5   H_TYPE         201 non-null    object 
 6   H_INDICATOR    201 non-null    object 
 7   GUEST_REGION   201 non-null    object 
 8   SOURCE_DETAIL  201 non-null    object 
 9   TIME_PERIOD    201 non-null    int64  
 10  OBS_VALUE      201 non-null    float64
dtypes: float64(1), int64(1), object(9)
memory usage: 17.4+ KB
None


- GDP

#### 4.11 `GDP_Quarterly_Constant_Prices.csv`

In [56]:
gdp_quarterly_constant_df = data_loader.inspect_data("11_gdp_quarterly_constant", delimiter=",", show_progress=False,loading_method='parallel')


Loading 11_gdp_quarterly_constant from ../datasets/raw/Gross Domestic Product/GDP_Quarterly_Constant_Prices.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.13it/s]

Shape of 11_gdp_quarterly_constant: (1920, 15)
Columns in 11_gdp_quarterly_constant:
Index(['DATAFLOW', 'REF_AREA', 'FREQ', 'UNIT_MEASURE', 'SOURCE_DETAIL',
       'MEASURE', 'QUARTER', 'QGDP_SYS', 'QGDP_UNIT', 'TIME_PERIOD',
       'OBS_VALUE', 'OBS_STATUS', 'UNIT_MULT', 'OBS_COMMENT', 'DECIMALS'],
      dtype='object')


In [57]:
gdp_quarterly_constant_df.isnull().mean() * 100

DATAFLOW           0.0
REF_AREA           0.0
FREQ               0.0
UNIT_MEASURE       0.0
SOURCE_DETAIL      0.0
MEASURE            0.0
QUARTER            0.0
QGDP_SYS           0.0
QGDP_UNIT          0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
OBS_STATUS       100.0
UNIT_MULT        100.0
OBS_COMMENT      100.0
DECIMALS           0.0
dtype: float64

In [58]:
processed_gdp_quarterly_constant_df = data_processor.format_and_clean_table(df=gdp_quarterly_constant_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1920 entries, 0 to 1919
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       1920 non-null   object 
 1   REF_AREA       1920 non-null   object 
 2   FREQ           1920 non-null   object 
 3   UNIT_MEASURE   1920 non-null   object 
 4   SOURCE_DETAIL  1920 non-null   object 
 5   MEASURE        1920 non-null   object 
 6   QUARTER        1920 non-null   object 
 7   QGDP_SYS       1920 non-null   object 
 8   QGDP_UNIT      1920 non-null   object 
 9   TIME_PERIOD    1920 non-null   int64  
 10  OBS_VALUE      1920 non-null   float64
 11  DECIMALS       1920 non-null   int64  
dtypes: float64(1), int64(2), object(9)
memory usage: 180.1+ KB
None


#### 4.12 `GDP_Quarterly_Current_Prices.csv`

In [59]:
gdp_quarterly_current_df = data_loader.inspect_data("12_gdp_quarterly_current", delimiter=",", show_progress=False,loading_method='parallel')


Loading 12_gdp_quarterly_current from ../datasets/raw/Gross Domestic Product/GDP_Quarterly_Current_Prices.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.12it/s]

Shape of 12_gdp_quarterly_current: (1920, 15)
Columns in 12_gdp_quarterly_current:
Index(['DATAFLOW', 'REF_AREA', 'FREQ', 'UNIT_MEASURE', 'SOURCE_DETAIL',
       'MEASURE', 'QUARTER', 'QGDP_SYS', 'QGDP_UNIT', 'TIME_PERIOD',
       'OBS_VALUE', 'OBS_STATUS', 'UNIT_MULT', 'OBS_COMMENT', 'DECIMALS'],
      dtype='object')


In [60]:
gdp_quarterly_current_df.isnull().mean() * 100

DATAFLOW           0.0
REF_AREA           0.0
FREQ               0.0
UNIT_MEASURE       0.0
SOURCE_DETAIL      0.0
MEASURE            0.0
QUARTER            0.0
QGDP_SYS           0.0
QGDP_UNIT          0.0
TIME_PERIOD        0.0
OBS_VALUE          0.0
OBS_STATUS       100.0
UNIT_MULT        100.0
OBS_COMMENT      100.0
DECIMALS           0.0
dtype: float64

In [61]:
processed_gdp_quarterly_current_df = data_processor.format_and_clean_table(df=gdp_quarterly_current_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1920 entries, 0 to 1919
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       1920 non-null   object 
 1   REF_AREA       1920 non-null   object 
 2   FREQ           1920 non-null   object 
 3   UNIT_MEASURE   1920 non-null   object 
 4   SOURCE_DETAIL  1920 non-null   object 
 5   MEASURE        1920 non-null   object 
 6   QUARTER        1920 non-null   object 
 7   QGDP_SYS       1920 non-null   object 
 8   QGDP_UNIT      1920 non-null   object 
 9   TIME_PERIOD    1920 non-null   int64  
 10  OBS_VALUE      1920 non-null   float64
 11  DECIMALS       1920 non-null   int64  
dtypes: float64(1), int64(2), object(9)
memory usage: 180.1+ KB
None


In [62]:
data_processor.display_dataframe_info(processed_gdp_quarterly_current_df)

Shape: (1920, 12)
Dtypes: ['DATAFLOW:object', 'REF_AREA:object', 'FREQ:object', 'UNIT_MEASURE:object', 'SOURCE_DETAIL:object', 'MEASURE:object', 'QUARTER:object', 'QGDP_SYS:object', 'QGDP_UNIT:object', 'TIME_PERIOD:int64', 'OBS_VALUE:float64', 'DECIMALS:int64']
Missings (%): ['DATAFLOW:0.00%', 'REF_AREA:0.00%', 'FREQ:0.00%', 'UNIT_MEASURE:0.00%', 'SOURCE_DETAIL:0.00%', 'MEASURE:0.00%', 'QUARTER:0.00%', 'QGDP_SYS:0.00%', 'QGDP_UNIT:0.00%', 'TIME_PERIOD:0.00%', 'OBS_VALUE:0.00%', 'DECIMALS:0.00%']
Total missings (%): 0.00%
total duplicates: 0
Total memory usage: 0.18 MB


DATAFLOW REF_AREA FREQ UNIT_MEASURE SOURCE_DETAIL MEASURE  \
0  FCSA:DF_QGDP_CUR(1.8.0)       AE    Q      PERCENT          FCSC     NFC   
1  FCSA:DF_QGDP_CUR(1.8.0)       AE    Q      PERCENT          FCSC     NFC   

  QUARTER QGDP_SYS  QGDP_UNIT  TIME_PERIOD  OBS_VALUE  DECIMALS  
0      Q3      CUR  GWTH_RATE         2013   4.355075         1  
1      Q3      CUR  GWTH_RATE         2014   4.394883         1

- World Development Indicators

#### 4.13 `Metadata_World_Development_Indicator.csv`

In [63]:
world_dev_metadata_df = data_loader.inspect_data("13_world_dev_metadata", delimiter=",", show_progress=False,loading_method='parallel')


Loading 13_world_dev_metadata from ../datasets/raw/World Development Indicators/Metadata_World_Development_Indicator.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.12it/s]

Shape of 13_world_dev_metadata: (1496, 5)
Columns in 13_world_dev_metadata:
Index(['INDICATOR_CODE', 'INDICATOR_NAME', 'SOURCE_NOTE',
       'SOURCE_ORGANIZATION', 'Unnamed: 4'],
      dtype='object')


In [64]:
world_dev_metadata_df.isnull().mean() * 100

INDICATOR_CODE           0.0
INDICATOR_NAME           0.0
SOURCE_NOTE              0.0
SOURCE_ORGANIZATION      0.0
Unnamed: 4             100.0
dtype: float64

In [65]:
processed_world_dev_metadata_df = data_processor.format_and_clean_table(df=world_dev_metadata_df, missing_threshold=0.8)


Formatted and Cleaned Data Overview:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1496 entries, 0 to 1495
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   INDICATOR_CODE       1496 non-null   object
 1   INDICATOR_NAME       1496 non-null   object
 2   SOURCE_NOTE          1496 non-null   object
 3   SOURCE_ORGANIZATION  1496 non-null   object
dtypes: object(4)
memory usage: 46.9+ KB
None


#### 4.14 `World_Development_Indicator.csv`

In [66]:
world_dev_indicators_df = data_loader.inspect_data("14_world_dev_indicators", delimiter=",", show_progress=False,loading_method='parallel', skiprows=4)


Loading 14_world_dev_indicators from ../datasets/raw/World Development Indicators/World_Development_Indicator.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:00<00:00,  1.13it/s]

Shape of 14_world_dev_indicators: (1496, 69)
Columns in 14_world_dev_indicators:
Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023', 'Unnamed: 68'],
      dtype='object')


In [67]:
world_dev_indicators_df

Country Name Country Code  \
0     United Arab Emirates          ARE   
1     United Arab Emirates          ARE   
2     United Arab Emirates          ARE   
3     United Arab Emirates          ARE   
4     United Arab Emirates          ARE   
...                    ...          ...   
1491  United Arab Emirates          ARE   
1492  United Arab Emirates          ARE   
1493  United Arab Emirates          ARE   
1494  United Arab Emirates          ARE   
1495  United Arab Emirates          ARE   

                                         Indicator Name  \
0                    International migrant stock, total   
1     Children in employment, wage workers (% of chi...   
2     Unemployment, total (% of total labor force) (...   
3     Unemployment, female (% of female labor force)...   
4     Share of youth not in education, employment or...   
...                                                 ...   
1491  Carbon dioxide (CO2) emissions excluding LULUC...   
1492  Carbon dioxide (CO2) net fluxes from LULUCF - ...   
1493  Carbon dioxide (CO2) emissions from Industrial...   
1494  Methane (CH4) emissions (total) excluding LULU...   
1495  Methane (CH4) emissions (total) excluding LULU...   

               Indicator Code    1960  1961  1962  1963  1964     1965  1966  \
0                 SM.POP.TOTL  2194.0   NaN   NaN   NaN   NaN  25981.0   NaN   
1              SL.WAG.0714.ZS     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
2           SL.UEM.TOTL.NE.ZS     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
3        SL.UEM.TOTL.FE.NE.ZS     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
4        SL.UEM.NEET.MA.ME.ZS     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
...                       ...     ...   ...   ...   ...   ...      ...   ...   
1491     EN.GHG.CO2.PC.CE.AR5     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
1492  EN.GHG.CO2.LU.MT.CE.AR5     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
1493  EN.GHG.CO2.IC.MT.CE.AR5     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
1494        EN.GHG.CH4.ZG.AR5     NaN   NaN   NaN   NaN   NaN      NaN   NaN   
1495     EN.GHG.CH4.MT.CE.AR5     NaN   NaN   NaN   NaN   NaN      NaN   NaN   

      1967  1968  1969          1970       1971       1972      1973  \
0      NaN   NaN   NaN  65827.000000        NaN        NaN       NaN   
1      NaN   NaN   NaN           NaN        NaN        NaN       NaN   
2      NaN   NaN   NaN           NaN        NaN        NaN       NaN   
3      NaN   NaN   NaN           NaN        NaN        NaN       NaN   
4      NaN   NaN   NaN           NaN        NaN        NaN       NaN   
...    ...   ...   ...           ...        ...        ...       ...   
1491   NaN   NaN   NaN     65.180285  73.710426  70.614831  81.06109   
1492   NaN   NaN   NaN           NaN        NaN        NaN       NaN   
1493   NaN   NaN   NaN      1.907600   1.907600   1.698800   1.93140   
1494   NaN   NaN   NaN           NaN        NaN        NaN       NaN   
1495   NaN   NaN   NaN      9.636400   9.690300  10.926200  13.83680   

           1974           1975       1976       1977       1978       1979  \
0           NaN  312400.000000        NaN        NaN        NaN        NaN   
1           NaN            NaN        NaN        NaN        NaN        NaN   
2           NaN       1.960000        NaN        NaN        NaN        NaN   
3           NaN       4.510000        NaN        NaN        NaN        NaN   
4           NaN            NaN        NaN        NaN        NaN        NaN   
...         ...            ...        ...        ...        ...        ...   
1491  73.084733      67.395849  74.108278  64.806518  65.260404  41.628837   
1492        NaN            NaN        NaN        NaN        NaN        NaN   
1493   1.665100       2.258800   2.760500   6.001600   6.546000   7.124100   
1494        NaN            NaN        NaN        NaN        NaN        NaN   
1495  15.079300      15.416700  17.645200  18.931700  18.319800  18.633100   

               1980  ... 

In [68]:
world_dev_indicators_melted_df = world_dev_indicators_df.melt(
    id_vars=["Country Name", "Country Code", "Indicator Name", "Indicator Code"],
    var_name="Year",
    value_name="Value"
)


In [69]:
world_dev_indicators_melted_df

Country Name Country Code  \
0      United Arab Emirates          ARE   
1      United Arab Emirates          ARE   
2      United Arab Emirates          ARE   
3      United Arab Emirates          ARE   
4      United Arab Emirates          ARE   
...                     ...          ...   
97235  United Arab Emirates          ARE   
97236  United Arab Emirates          ARE   
97237  United Arab Emirates          ARE   
97238  United Arab Emirates          ARE   
97239  United Arab Emirates          ARE   

                                          Indicator Name  \
0                     International migrant stock, total   
1      Children in employment, wage workers (% of chi...   
2      Unemployment, total (% of total labor force) (...   
3      Unemployment, female (% of female labor force)...   
4      Share of youth not in education, employment or...   
...                                                  ...   
97235  Carbon dioxide (CO2) emissions excluding LULUC...   
97236  Carbon dioxide (CO2) net fluxes from LULUCF - ...   
97237  Carbon dioxide (CO2) emissions from Industrial...   
97238  Methane (CH4) emissions (total) excluding LULU...   
97239  Methane (CH4) emissions (total) excluding LULU...   

                Indicator Code         Year   Value  
0                  SM.POP.TOTL         1960  2194.0  
1               SL.WAG.0714.ZS         1960     NaN  
2            SL.UEM.TOTL.NE.ZS         1960     NaN  
3         SL.UEM.TOTL.FE.NE.ZS         1960     NaN  
4         SL.UEM.NEET.MA.ME.ZS         1960     NaN  
...                        ...          ...     ...  
97235     EN.GHG.CO2.PC.CE.AR5  Unnamed: 68     NaN  
97236  EN.GHG.CO2.LU.MT.CE.AR5  Unnamed: 68     NaN  
97237  EN.GHG.CO2.IC.MT.CE.AR5  Unnamed: 68     NaN  
97238        EN.GHG.CH4.ZG.AR5  Unnamed: 68     NaN  
97239     EN.GHG.CH4.MT.CE.AR5  Unnamed: 68     NaN  

[97240 rows x 6 columns]

In [70]:
world_dev_indicators_melted_df.isnull().mean() * 100

Country Name       0.000000
Country Code       0.000000
Indicator Name     0.000000
Indicator Code     0.000000
Year               0.000000
Value             69.714109
dtype: float64

- Currency Strength

#### 4.15 `AED-USD.csv`

In [72]:
currency_strength_df = data_loader.inspect_data("15_currency", delimiter=",", show_progress=False,loading_method='parallel')


Loading 15_currency from ../datasets/raw/Currency Strength/AED-USD.csv...


Loading files in parallel: 100%|██████████| 1/1 [00:01<00:00,  1.17s/it]

Shape of 15_currency: (5468, 9)
Columns in 15_currency:
Index(['Unnamed: 0', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume',
       'Date', 'Return'],
      dtype='object')


In [73]:
currency_strength_df

Unnamed: 0      Open      High       Low     Close  Adj Close  Volume  \
0              0  0.272331  0.272368  0.272331  0.272368   0.272368       0   
1              1  0.272413  0.272420  0.272331  0.272368   0.272368       0   
2              2  0.272339  0.272368  0.272331  0.272368   0.272368       0   
3              3  0.272346  0.272368  0.272331  0.272368   0.272368       0   
4              4  0.272339  0.272346  0.272331  0.272346   0.272346       0   
...          ...       ...       ...       ...       ...        ...     ...   
5463        5463  0.272331  0.272331  0.272264  0.272331   0.272331       0   
5464        5464  0.272331  0.272331  0.272264  0.272331   0.272331       0   
5465        5465  0.272331  0.272331  0.272270  0.272331   0.272331       0   
5466        5466  0.272331  0.272331  0.272264  0.272331   0.272331       0   
5467        5467  0.272272  0.272280  0.272271  0.272279   0.272279       0   

            Date    Return  
0     2003-12-01       NaN  
1     2003-12-02  0.000000  
2     2003-12-03  0.000000  
3     2003-12-04  0.000000  
4     2003-12-05 -0.008174  
...          ...       ...  
5463  2024-11-26  0.019056  
5464  2024-11-27  0.000000  
5465  2024-11-28  0.000000  
5466  2024-11-29  0.000000  
5467  2024-12-02 -0.019052  

[5468 rows x 9 columns]

In [74]:
currency_strength_df.isnull().mean() * 100

Unnamed: 0    0.000000
Open          0.000000
High          0.000000
Low           0.000000
Close         0.000000
Adj Close     0.000000
Volume        0.000000
Date          0.000000
Return        0.018288
dtype: float64

### **5. ` Initial Cleaning`**

In [94]:
# Perform initial cleaning
# rents_cleaned = clean_data(rents_data)

# Inspect cleaned data
currency_strength_df.drop(columns='Unnamed: 0', inplace=True)
currency_strength_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5468 entries, 0 to 5467
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       5468 non-null   float64
 1   High       5468 non-null   float64
 2   Low        5468 non-null   float64
 3   Close      5468 non-null   float64
 4   Adj Close  5468 non-null   float64
 5   Volume     5468 non-null   int64  
 6   Date       5468 non-null   object 
 7   Return     5467 non-null   float64
dtypes: float64(6), int64(1), object(1)
memory usage: 341.9+ KB


### **5. `  Save Cleaned Data`**

In [ ]:
# Save cleaned data to the processed folder
# processed_rents_file = os.path.join(processed_data_path, "real_estate_rents_cleaned.csv")
# rents_cleaned.to_csv(processed_rents_file, index=False)

# print(f"Cleaned rents dataset saved to {processed_rents_file}")


In [91]:
pop_indicators_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   DATAFLOW       18 non-null     object 
 1   MEASURE        18 non-null     object 
 2   POP_IND        18 non-null     object 
 3   UNIT_MEASURE   18 non-null     object 
 4   REF_AREA       18 non-null     object 
 5   FREQ           18 non-null     object 
 6   GENDER         18 non-null     object 
 7   SOURCE_DETAIL  18 non-null     object 
 8   TIME_PERIOD    18 non-null     int64  
 9   OBS_VALUE      18 non-null     float64
 10  DECIMALS       18 non-null     int64  
dtypes: float64(1), int64(2), object(8)
memory usage: 1.7+ KB


In [81]:
processed_rents_df.to_parquet("../datasets/processed/processed_rents_df.parquet")
processed_transactions_df.to_parquet("../datasets/processed/processed_transactions_df.parquet")

In [84]:
processed_gdp_quarterly_constant_df.to_parquet("../datasets/processed/processed_gdp_quarterly_constant_df.parquet")
processed_gdp_quarterly_current_df.to_parquet("../datasets/processed/processed_gdp_quarterly_current_df.parquet")


In [88]:
processed_tourism_indicators_df.to_parquet("../datasets/processed/processed_tourism_indicators_df.parquet")
processed_tourism_regions_df.to_parquet("../datasets/processed/processed_tourism_regions_df.parquet")
processed_tourism_ratings_df.to_parquet("../datasets/processed/processed_tourism_ratings_df.parquet")

In [95]:
processed_pop_estimates_df.to_parquet("../datasets/processed/processed_pop_estimates_df.parquet")
pop_indicators_df.to_parquet("../datasets/processed/pop_indicators.parquet")
currency_strength_df.to_parquet("../datasets/processed/currency_strength_df.parquet")
